Banking Assistant Documentation (#151)
cisco/mindmeld: source/blueprints/banking.rst

Banking Assistant
=================

In this step-by-step walkthrough, you will build a conversational application that allows users to complete common banking tasks that include transferring money and paying bills.

Working through this blueprint will teach you how to

   - obtain missing information (entities) using a slot/entity filling form
   - query and update external data storage with the dialogue manager
   - pass immutable information, like a user token, from a client application to your MindMeld application server
   - use roles in entity hierarchy

.. note::

   Please make sure to install and run all of the :ref:`pre-requisites <getting_started_virtualenv_setup>` for MindMeld before continuing on with this blueprint tutorial.

1. The Use Case
^^^^^^^^^^^^^^^

> Users should be able to securely access their banking information and complete tasks as if they are conversing with a teller. They should be able to check balances, pay off credit card debt, transfer money, and carry out various other banking operations they may perform on a day to day basis. In a production application, the frontend would likely handle the secure user authentication and pass a token to the MindMeld server, which would allow the application to make calls to a bank's REST APIs.

Reviewer: `the frontend would likely handle the secure user authentication and pass a token to the MindMeld server` -> `the frontend would likely handle user authentication and pass an authenticated token to the MindMeld server`

> Users should be able to securely access their banking information and complete tasks as if they are conversing with a teller. They should be able to check balances, pay off credit card debt, transfer money, and carry out various other banking operations they may perform on a day to day basis. In a production application, the frontend would likely handle the secure user authentication and pass a token to the MindMeld server, which would allow the application to make calls to a bank's REST APIs.

Reviewer: ```suggestion
Users should be able to securely access their banking information and complete tasks as if they are conversing with a teller. They should be able to check balances, pay off credit card debt, transfer money, and carry out various other banking operations they may perform on a day to day basis. In a production application, the frontend would likely handle the secure user authentication and pass a token to the MindMeld server, which would allow the application to make calls to a bank's REST APIs.
```

> Users should be able to securely access their banking information and complete tasks as if they are conversing with a teller. They should be able to check balances, pay off credit card debt, transfer money, and carry out various other banking operations they may perform on a day to day basis. In a production application, the frontend would likely handle the secure user authentication and pass a token to the MindMeld server, which would allow the application to make calls to a bank's REST APIs.

Reviewer: ```suggestion
Users should be able to securely access their banking information and complete tasks as if they are conversing with a teller. They should be able to check balances, pay off credit card debt, transfer money, and carry out various other banking operations they may perform on a day to day basis. In a production application, the frontend would handle user authentication and pass a token to the MindMeld server, which would allow the application to make calls to a bank's REST APIs.
```

2. Example Dialogue Interactions
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The conversational flows for a banking assistant can be highly complex, depending on the desired application functionality and the amount of user guidance required at each step. Anticipating all user interactions requires multiple iterations. 

While a real production application will securely authenticate users, for this blueprint, we have given the option to choose from three sample users. You can specify the desired username as a parameter in the request context and if a user is not specified a random user from the sample users will be chosen.

Sample usernames: ``splashbro30``, ``larry_l24``, ``johndoe123``

.. code-block:: console

>   conv = Conversation(nlp=nlp, app_path='banking_assistant', context={'user_name' : 'splashbro30'})

Reviewer: ```suggestion
  conv = Conversation(nlp=nlp, app_path='banking_assistant', context={'user_name' : 'alice'})
```

>   conv = Conversation(nlp=nlp, app_path='banking_assistant', context={'user_name' : 'splashbro30'})

Reviewer: Can we change the username to something more conventional, like `Alice` or `Bob`?

3. Domain-Intent-Entity Hierarchy
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The banking assistant is organized into four domains: ``Greeting``, ``Accounts & Credit Cards``, ``FAQ``, and ``Unsupported``. The main domain, ``accounts_creditcards``, encompasses all of the functionality required to perform tasks that require personal user banking information. The ``accounts_creditcards`` domain supports the following intents:

   - ``transfer_balances`` — User wants to transfer money from one account to another
   - ``check_balances`` — User wants the current balance for a particular account
   - ``pay_creditcard`` — User wants to pay off a credit card bill
   - ``check_due_date`` — User wants the date the bill is due 
   - ``setup_autopay`` — User wants to activate automatic payments

The full list of intents for all domains is illustrated below.

.. image:: /images/banking_hierarchy.png

> There are two types of entities in MindMeld: :ref:`System Entities <system-entities>` and :doc:`Custom Entities <../userguide/entity_recognizer>`. System entities are pre-defined and automatically detected by MindMeld. Examples include sys_amount-of-money, sys_time, and sys_interval. Custom entities are defined by the developers of each application and detected by the models trained on the labeled data and entity gazetteer files.

Reviewer 2: Minor nit
```suggestion
There are two types of entities in MindMeld: :ref:`System Entities <system-entities>` and :doc:`Custom Entities <../userguide/entity_recognizer>`. System entities are pre-defined and automatically detected by MindMeld. Examples include ``sys_amount-of-money``, ``sys_time``, and ``sys_interval``. Custom entities are defined by the developers of each application and detected by the models trained on the labeled data and entity gazetteer files.
```

Banking Assistant
=================

> In this step-by-step walkthrough, you will use a MindMeld blueprint to build a conversational application that allows users to complete common banking tasks that include transferring money and paying bills.

Reviewer 3: ```suggestion
In this step-by-step walkthrough, you will build a conversational application that allows users to complete common banking tasks that include transferring money and paying bills.
```

.. note::

   Please make sure to install and run all of the :ref:`pre-requisites <getting_started_virtualenv_setup>` for MindMeld before continuing on with this blueprint tutorial.

1. The Use Case
^^^^^^^^^^^^^^^

> Users should be able to securely access their banking information and complete tasks as if they are conversing with a teller. They should be able to check balances, pay off credit card debt, transfer money, and carry out various other banking operations they may perform on day to day basis. In a production application, the frontend would likely handle the secure user authentication and pass a token to the MindMeld server, which would allow the application to make calls to a bank's REST API.

Reviewer 3: ```suggestion
Users should be able to securely access their banking information and complete tasks as if they are conversing with a teller. They should be able to check balances, pay off credit card debt, transfer money, and carry out various other banking operations they may perform on a day to day basis. In a production application, the frontend would likely handle the secure user authentication and pass a token to the MindMeld server, which would allow the application to make calls to a bank's REST APIs.
```

2. Example Dialogue Interactions
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> The conversational flows for the banking assistant can be highly complex, depending on the desired application functionality and the amount of user guidance required at each step. Anticipating all user interactions requires multiple iterations. 

Reviewer 3: ```suggestion
The conversational flows for a banking assistant can be highly complex, depending on the desired application functionality and the amount of user guidance required at each step. Anticipating all user interactions requires multiple iterations. 
```

> When using the application, we have given the option to choose three sample users. You can provide the specified user you would like to interact as a parameter in the request context and if a user is not specified a random user from the sample users will be chosen. 

Reviewer 3: ```suggestion
While a real production application will securely authenticate users, for this blueprint, we have given the option to choose from three sample users. You can specify the desired username as a parameter in the request context and if a user is not specified a random user from the sample users will be chosen.
```

> Sample User Usernames: ``splashbro30``, ``larry_l24``, ``johndoe123``

Reviewer 3: ```suggestion
Sample usernames: ``splashbro30``, ``larry_l24``, ``johndoe123``
```

The full list of intents for all domains is illustrated below.

.. image:: /images/banking_hierarchy.png

There are two types of entities in MindMeld: :ref:System Entities <system-entities> and :doc:Custom Entities <../userguide/entity_recognizer>. System entities are pre-defined and automatically detected by MindMeld. Examples include sys_amount-of-money, sys_time, and sys_interval. Custom entities are defined by the developers of each application and detected by the models trained on the labeled data and entity gazetteer files.

The banking assistant defines and uses the following custom entities in the accounts_creditcards domain:

    - ``account_type``: detects which account you are referring to, for example: "what is my {savings|account_type} balance?"
    - ``credit_amount``: detects the amount you would like to pay off your credit bill, for example: "can you make a {minimum|credit_amount} payment for my card?"

The banking assistant uses one system entity ``sys_amount-of-money`` (money). Some examples for annotation with system entities: "please transfer {20k|sys_amount-of-money}" and "pay {two hundred dollars|sys_amount-of-money} towards my credit bill".

> Queries can include more than one entity of the same type. In "transfer money from checking to savings.", for example, both "checking" and "savings" are ``account_type`` entities. Just labeling both entities with the same type does not give MindMeld enough information to understand the meaning of the query. We need to show the different *roles* that these two ``account_type`` entities play in creating meaning. One is an "origin" that the money is coming from, and the other is a "destination" where the user wants to move the money to. We annotate the example as "transfer money from {checking|account_type|origin} to {savings|account_type|dest}." This way, MindMeld can interpret each entity correctly. See :doc:`Role Classifier <../userguide/role_classifier>`.

Reviewer 3: ```suggestion
Queries can include more than one entity of the same type. In "transfer money from checking to savings", for example, both "checking" and "savings" are ``account_type`` entities. Simply detecting these two entities as ``account_type`` doesn't provide us with all the information we need to fully understand the user query and execute the desired action. We therefore use *roles* to further classify the ``account_type`` entities - an "origin" role that represents where the money is coming from, and a "destination" role that indicates where the money needs to be moved to. Accordingly, we annotate the example as "transfer money from {checking|account_type|origin} to {savings|account_type|dest}." This way, we can interpret each entity correctly and better understand the meaning of the full query. See :doc:`Role Classifier <../userguide/role_classifier>`.
```

		    #update the checking account to reflect the payment made 
		    _put('checking', get('checking') - responder.slots['min'])
		#if it is balance handle it similar to minimum but with the amount being the total credit debt
                else:
                    responder.reply(['Sure, we have scheduled your credit card payment for your {payment} of ${total_balance}.'])
                    _put('credit', 0)
		    _put('checking', get('checking') - responder.slots['total_balance'])
	    #handle the system entity of the exact dollar amount they want to pay off
            else:
                responder.slots['amount'] = entity['value'][0]['value']
                responder.reply(['Ok we have scheduled your credit card payment for {amount}'])
                _put('credit', _get('credit') - entity['value'][0]['value'])
		_put('checking', get('checking') - responder.slots['amount'])
	responder.listen() 

For a more realistic implementation of ``pay_creditcard`` that deals with varied user flows and the full code behind all the dialogue state handlers, see the ``__init__.py`` file in the blueprint folder.

> 5. Slot Filling

Reviewer 3: @abhijeetsidhu Could make this 4.1 or just an non-numbered sub-section under **4. Dialogue States**? We want to stick to 10 total sections since they have a one-to-one correspondence with 10-step guide in the documentation.

Author: Sure will make it a non-numbered sub-section.

^^^^^^^^^^^^^^^

> Slot/entity filling logic allows you to easily request for a missing entity. You can set custom responses in the slot filling form to prompt the user with when an entity is missing in the user request. Once the missing entities in the form have been provided you can handle the rest of the logic as you would in the handler function of your intent. More detailed information on slot filling is below along with an example from the banking assistant. 

Reviewer 3: ```suggestion
Slot/entity filling logic allows you to easily request for a missing entity. You can set custom responses in the slot filling form to prompt the user with when an entity is missing in the user request. Once the missing entities in the form have been provided you can handle the rest of the logic as you would in the handler function of your intent.
```

  - ``max_retries`` - (int, optional, default 1): maximum number of retries allowed per entity or slot if user response is invalid.
  - ``exit_msg`` - (str, optional): If slot filling is exited abruptly without completion, define custom message to display.
  - ``exit_keys`` - (list, optional): List of exit hints for the slot filling flow. If these words or phrases are said by the user, the slot filling logic exits. Default: ['cancel', 'restart', 'exit', 'reset'].

FormEntity is a class that allows creation of entity objects for slot filling and comprises of the following attributes:

  - ``entity``  - (str, required): Entity name
  - ``role`` - (str, optional): The role of the entity
  - ``responses`` - (list or str, optional): message for prompting the user for missing entities
  - ``retry_response`` - (str, optional): message for re-prompting users. If not provided, defaults to responses.
  - ``value`` - (str, optional): The resolved value of the entity
  - ``default_eval`` - (bool, optional): Use system validation (default: True)
  - ``hints`` - (list, optional): Developer defined list of keywords to verify theuser input against

>   - ``custom_eval`` - (func, optional): custom validation function (should return bool:validated or not)

Reviewer 3: @abhijeetsidhu  @ritvikshrivastava I don't think we need to repeat the documentation that's already in the user guide. It would be better to just link to the appropriate section. Restrict the content in this section to just the code example from the banking blueprint.

                'new task', 'other', 'return'],
	#a message to prompt the user after exit
        'exit_msg' : 'A few other banking tasks you can try are, ordering checks and paying bills',
	#the number of max tries for the user to specify the entity
	'max_retries' = 1
    }

    #the app auto fill indicates it is a dialogue state which requires a form and uses slot filling logic	
    @app.auto_fill(intent='check_balances', form=balance_form)
    #the function will work as if the required entity has been provided 
    def check_balances_handler(request, responder):
        if not user_data: 
            _pull_data(request)
        if request.entities:
            for entity in request.entities:
                if entity['type'] == 'account_type':
                    responder.slots['account'] = entity['value'][0]['cname'] or entity['text']
                    responder.slots['amount'] = _get(entity['value'][0]['cname'] or entity['text'])
                    responder.reply('Your {account} account balance is {amount}')
        

6. Knowledge Base
^^^^^^^^^^^^^^^^^

> Since the banking assistant will require personal information we have decided to not include a knowledge base to mimic how this data may be received from an

Reviewer 2: broken line?

> external api. Most likely the frontend would handle the user authenication and pass on a token to the request context, which is immutable once the conversation has started. The MindMeld server can then use this token to retrieve user information from a secure database. In this blueprint we demonstrate how this can be done by simply passing the user_name in the context and using it to retrieve information from a simple database. For a production app, these can be replaced by a secure token and an API call to a secure database. These databases and APIs likely already exist internally at bank organizations.

Reviewer 2: ```suggestion
external API. Most likely the frontend would handle the user authentication and pass on a token to the request context, which is immutable once the conversation has started. The MindMeld server can then use this token to retrieve user information from a secure database. In this blueprint, we demonstrate how this can be done by simply passing the `user_name` in the context and using it to retrieve information from a simple database. For a production application, these can be replaced by a secure token and an API call to a secure database. These databases and APIs likely already exist internally at bank organizations.
```

> external API. Most likely the frontend would handle the user authentication and pass on a token to the request context, which is immutable once the conversation has started. The MindMeld server can then use this token to retrieve user information from a secure database. In this blueprint, we demonstrate how this can be done by simply passing the `user_name` in the context and using it to retrieve information from a simple database. For a production application, these can be replaced by a secure token and an API call to a secure database. These databases and APIs likely already exist internally at bank organizations.

Reviewer 3: ```suggestion
external API. Most likely the frontend would handle the user authentication and pass on a token to the request context, which is immutable once the conversation has started. The MindMeld server can then use this token to retrieve user information from a secure database. In this blueprint, we demonstrate how this can be done by simply passing the `user_name` in the context and using it to retrieve information from a simple database. For a production application, these can be replaced by a secure token and an API call to a secure database. These databases and APIs likely already exist internally at financial institutions.
```

> external API. Most likely the frontend would handle the user authentication and pass on a token to the request context, which is immutable once the conversation has started. The MindMeld server can then use this token to retrieve user information from a secure database. In this blueprint, we demonstrate how this can be done by simply passing the `user_name` in the context and using it to retrieve information from a simple database. For a production application, these can be replaced by a secure token and an API call to a secure database. These databases and APIs likely already exist internally at bank organizations.

Reviewer 3: ```suggestion
external API. Most likely the frontend would handle the user authentication and pass on a token to the request context, which is immutable once the conversation has started. The MindMeld server can then use this token to retrieve user information from a secure database. In this blueprint, we demonstrate how this can be done by simply passing the `user_name` in the context and using it to retrieve information from a simple database. For a production application, these can be replaced by a secure token and an API call to a secure database. These databases and APIs likely already exist internally at financial institutions.
```

| | training the Role Classifier       | | corresponding roles, e.g. ``origin`` and ``dest``."                       |
| |                                    | |                                                                           |
+--------------------------------------+-----------------------------------------------------------------------------+

The ``domains`` directory contains the training data for intent classification and entity recognition. The ``entities`` directory contains the data for entity resolution. Both directories are at root level in the blueprint folder.

.. admonition:: Exercise 1

   - Study the best practices around training data generation and annotation for conversational apps in :doc:`Step 6 <../quickstart/06_generate_representative_training_data>` of the Step-By-Step Guide. Following those principles, create additional labeled data for all the intents in this blueprint. Read more about :doc:`NLP model evaluation and error analysis <../userguide/nlp>` in the User Guide. Then apply what you have learned in evaluating your app, using your newly-created labeled data as held-out validation data.

>    - Complete the following exercise if you are extending the blueprint to build your own banking assistant. For app-agnostic, generic intents like ``greet``, ``exit``, and ``help``, start by simply reusing the blueprint data to train NLP models for your banking assistant. For ``pay_creditcard`` and any other app-specific intents, gather new training data tailored to the relevant entities (credit card, balance, etc.). Apply the approach you learned in :doc:`Step 6 <../quickstart/06_generate_representative_training_data>`.

Reviewer 2: ```suggestion
   - Complete the following exercise if you are extending the blueprint to build your own banking assistant. For common intents like ``greet``, ``exit``, and ``help``, start by simply reusing the blueprint data to train NLP models for your banking assistant. For ``pay_creditcard`` and any other app-specific intents, gather new training data tailored to the relevant entities (credit card, balance, etc.). Apply the approach you learned in :doc:`Step 6 <../quickstart/06_generate_representative_training_data>`.
```

   Loading queries from file banking_assistant/domains/faq/activate_creditcard/train.txt
   Fitting intent classifier: domain='accounts_creditcards'
   Selecting hyperparameters using k-fold cross-validation with 5 splits
   Best accuracy: 95.92%, params: {'C': 10, 'class_weight': {0: 0.8992592592592593, 1: 1.2360330578512397, 2: 0.8992592592592593, 3: 0.8033777777777779, 4:              
   1.6324705882352941}, 'fit_intercept': False}
   Fitting entity recognizer: domain='accounts_creditcards', intent='transfer_balances'
   Selecting hyperparameters using k-fold cross-validation with 5 splits
   Best accuracy: 97.70%, params: {'C': 100, 'penalty': 'l2'}
   Fitting intent classifier: domain='faq'
   Selecting hyperparameters using k-fold cross-validation with 5 splits
   Best accuracy: 95.76%, params: {'C': 100, 'class_weight': {0: 1.140406976744186, 1: 0.8602713178294573, 2: 1.0528645833333332, 3: 0.9125, 4: 0.9630733944954128, 5:     
   1.1307471264367814, 6: 0.9230603448275863, 7: 1.150294117647059}, 'fit_intercept': True}
   Fitting entity recognizer: domain='faq', intent='forgot_pin'

.. tip::

>   During active development, it's helpful to increase the :doc:`MindMeld logging level <../userguide/getting_started>` to better understand what's happening behind the scenes. All code snippets here assume that logging level has been set to verbose.

Reviewer 2: ```suggestion
  During active development, it is helpful to increase the :doc:`MindMeld logging level <../userguide/getting_started>` to better understand what is happening behind the scenes. All code snippets here assume that the logging level has been set to verbose.
```

.. code:: python

   features = {
               'bag-of-words': {'lengths': [1, 2]},
               'freq': {'bins': 5},
               'in-gaz': {},
               'length': {}
              }
   ic.fit(features=features)

.. code-block:: console

    Fitting intent classifier: domain='accounts_creditcards'
    Selecting hyperparameters using k-fold cross-validation with 5 splits
    Best accuracy: 95.92%, params: {'C': 10, 'class_weight': {0: 0.8992592592592593, 1: 1.2360330578512397, 2: 0.8992592592592593, 3: 0.8033777777777779, 4: 1.6324705882352941}, 'fit_intercept': False}

Change the classification model to random forest instead of the default logistic regression:

.. code:: python

   ic.fit(model_settings={'classifier_type': 'rforest'}, param_selection={'type': 'k-fold', 'k': 10, 'grid': {'class_bias': [0.7, 0.3, 0]}})

.. code-block:: console

    Fitting intent classifier: domain='accounts_creditcards'
    Selecting hyperparameters using k-fold cross-validation with 10 splits
    Best accuracy: 92.46%, params: {'class_weight': {0: 1.0, 1: 1.0, 2: 1.0, 3: 1.0, 4: 1.0}}

> You can use similar options to inspect and experiment with the Entity Recognizer and the other NLP classifiers. Finding the optimal machine learning settings is a highly iterative process involving several rounds of model training (with varying configurations), testing, and error analysis. See the User Guide for more about training, tuning, and evaluating the various MindMeld classifiers.

Reviewer 2: Add the link to User Guide

9. Parser Configuration
^^^^^^^^^^^^^^^^^^^^^^^

The relationships between entities in the banking assistant queries are simple ones. For example, in the annotated query ``can you pay my {minimum|credit_amount} bill``, the ``credit amount`` entity is self-sufficient, in that it is not described by any other entity.

If you extended the app to support queries with more complex entity relationships, it would be necessary to specify *entity groups* and configure the parser accordingly. For example, in the query ``can you pay my {minimum|credit_amount} bill for my {amex|credit_card}``, we would need to relate the ``credit_amount`` entity to the ``credit_card`` entity, because one entity describes the other. The related entities would form an entity group. For more about entity groups and parser configurations, see the :doc:`Language Parser <../userguide/parser>` chapter of the User Guide.

Since we do not have entity groups in the banking assistant, we do not need a parser configuration.

10. Using the Question Answerer
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> The :doc:`Question Answerer <../userguide/kb>` component in MindMeld is mainly used within dialogue state handlers for retrieving information from the knowledge base. Since the banking assistant has no knowledge base, no question answerer is not needed.

Reviewer 3: ```suggestion
The :doc:`Question Answerer <../userguide/kb>` component in MindMeld is mainly used within dialogue state handlers for retrieving information from the knowledge base. Since the banking assistant has no knowledge base, question answerer is not needed.
```

11. Testing and Deployment
^^^^^^^^^^^^^^^^^^^^^^^^^^

Once all the individual pieces (NLP, Question Answererer, Dialogue State Handlers) have been trained, configured, or implemented, use the :class:`Conversation` class in MindMeld to perform an end-to-end test of your conversational app.

For instance:

.. code:: python

   from mindmeld.components.dialogue import Conversation
   conv = Conversation(nlp=nlp, app_path='banking_assistant')
   conv.say("pay off my credit card bill for my remaining balance please")

.. code-block:: console

   ['Ok we have scheduled your credit card payment for your balance of $5000']

> The :meth:`say()` method packages the input text in a user request object and passes it to the MindMeld Application Manager to simulate a user interacting with the application. The method then outputs the textual part of the response sent by the app's Dialogue Manager. In the above example, we requested to pay off our credit debt, in a single query. The app responded, as expected, with the amount that will be paid off.

Reviewer 2: ```suggestion
The :meth:`say()` method packages the input text in a user request object and passes it to the MindMeld Application Manager to simulate a user interacting with the application. The method then outputs the textual part of the response sent by the application's Dialogue Manager. In the above example, we requested to pay off our credit debt, in a single query. The app responded, as expected, with the amount that will be paid off.
```

> You can also try out multi-turn dialogues(transfer balances):

Reviewer 2: ```suggestion
You can also try out multi-turn dialogues (transfer balances):
```

.. code:: python

>    >>> conv = Conversation(nlp=nlp, app_path='banking_assistant')

Reviewer 2: You can also pass the username in the context

Reviewer 3: Yes, would be good to show an example with the username passed into the context since that's one of the new things in this blueprint.
